### ISSUE TITLE
capscale: deprecate, downgrade or upgrade?

### ISSUE THREAD

**Issue by jarioksa** [vegandevs/vegan#217]
I'm re-implementing constrained ordination methods using common underlying code `ordConstrained()` in branch [rda-2.0](https://github.com/vegandevs/vegan/tree/rda-2.0). I have already implemented and merged to master `cca`, `rda` and `capscale` and I'm ready to start finishing `dbrda`. The `dbrda` re-implementation will break some support functions (`permutest` and hence `anova`, `fitted` and hence `residuals`, `stressplot`), and fixing those will break `capscale` at least for `permutest` and `anova`. There are at least two alternative ways of fixing the `capscale` breakage: upgrade to have special handling of  `capscale` to give similar results as in **vegan** release 2.4-0 (released Jun 15, 2016), or downgrade to have similar results as in older  releases (2.3-5 and older).

What happened in release 2.4-0 was that we added `dbrda` as a new function for distance-based RDA. This function directly handles dissimilarities and also handles negative eigenvalues. The old `capscale` (since **vegan** 1.6-0 of 2003) only analysed the real components and only gave information of the magnitude of non-analysed imaginary components (sum of negative eigenvalues). When we added `dbrda` we also made `capscale` to evaluate the magnitudes of partial, constrained and residual components with these negative eigenvalues although the actual ordination was only based on real components with positive eigenvalues. In this way, `anova` etc of `capscale` gave identical results to `dbrda`, although the ordination (plots) only used real components. I have no intention of implementing this behaviour in the new unified `ordConstrained` infrastructure, but currently `capscale` function adds the necessary components to its result, and current `dbrda` uses `capscale`-like internal structure. The new code will change `dbrda` to use similar structure as `cca` and `rda`, and I will remove the special code for distance-based methods in the support functions like `permutests.cca` (hence the breakage). Now the question is whether I still should keep that code only for `capscale` (i.e., upgrade), or remove the special handling from the support functions and parallel analysis from `capscale` (i.e., to downgrade).

Finally, the big question is do we really need `capscale` any longer? It has been in **vegan** for thirteen years, but it is superseded by `dbrda`. We could deprecate `capscale` in release `2.4-2` and make defunct in 2.6-0 (i.e., not yet in 2.5-0 where it would be deprecated). Should it be downgraded before that? Or should we keep  downgraded `capscale` only handling real dimensions & positive eigenvalues?

**Comment by EDiLD** [comment]
My 2cts:
Let `capscale` go and keep `dbrda` with support functions...

**Comment by jarioksa** [comment]
I have now downgraded `capscale` to its pre-2.4-0 shape before introducing `dbrda`. The function is not deprecated  yet, though, and the changes are not yet merged to the main rda-2.0 branch. Probably some more testing will be needed.

**Comment by jarioksa** [comment]
Downgraded `capscale` is now merged to the rda-2.0 branch (commit  05080f8) and once rda-2.0 is merged to master, the downgraded version will be there. This merge will happen after making `dbrda` to build upon `ordConstrained`. 

There is no deprecation message yet, and I haven't made my mind of deprecating `capscale`, although I probably will do so finally.

**Comment by gavinsimpson** [comment]
Is there any harm in leaving this function as is? We break code as soon as we make it Defunct and I'm not convinced there is anything fundamentally wrong with the `capscale` method that would warrant its removal.

**Comment by jarioksa** [comment]
I have decided to keep this function like @gavinsimpson suggested. There is some maintenance cost, but we have worse methods in **vegan** (`anosim`, `rrarefy`, `decorana`,...)

**Comment by jarioksa** [comment]
After discussion in #636 I decided to resurrect the issue of deprecating `capscale`.  I started with comparing these two functions to see if they offer similar support so that no major functionality is lost in deprecation.  With Euclidean distances,  `rda`, `capscale` and `dbrda` give same eigenvalues and the same row scores,  both LC and WA (with possible sign reversals), and with Euclidean distances all these can be regarded as the same. There are two major differences that will be discussed below:
- `dbrda` does not return species scores, but `capscale` does.
- original input data cannot be reconstructed from the ordination results in `dbrda`, but it is possible in `capscale` (although that is not the input user supplied).

`dbrda` does not return species scores even if rectangular community data are supplied, whereas `capscale` does return those. The species scores in `capscale` with Euclidean distance are same (with possible sign reversals as in `rda`). This was a design decision in `dbrda`: while these species scores are correct with Euclidean distances, they may not be adequate with other distances. Even with distances that can be expressed as Euclidean with data transformation (Hellinger, Chord _etc_) they would be based on non-transformed data, and with non-Euclidean dissimilarities (Bray-Curtis and all that) it is dubious to use Euclidean projection of original data. However, I provided `sppscores<-` functions to add species scores allowing transformation of the input data. Species scores must be actively added and this may disturb users when `capscale` adds scores automatically (even if wrong). 

There is one inevitable difference due to algorithms and it has consequences on several methods: in `dbrda` we cannot easily decompose data into additive components (although variance can be decomposed). The internal data structures can be accessed with `ordiYbar` function. In `rda` and `capscale` these are similar (even if not identical), and they will decompose the variation into additive partial, constrained and residual components. Let **Y** be the rectangular input matrix, **X** the constrained matrix and **Z** residual unconstrained matrix. These can be extracted as `ordiYbar` models `"initial"`, `"CCA"` and `"CA"`. In `rda` and `capscale` the original (internal) data can be reconstructed as **Y** = **X** + **Z**.   In `dbrda` the internal components are Gower transformed double-centred dissimilarities, but these are identical to outer products of `rda` and `capscale`, so that the `"initial"` input data is **YY'**. These components are not additive: the relationship is **YY'** = (**X** + **Z**)(**X** + **Z**)**'** = **XX'** + **ZZ'** + **XZ'** + **ZX'**. Matrices **YY'**, **XX'** and **ZZ'** are returned with `ordiYbar`, but **XZ'** and **ZX'** (which are transposes of each other) cannot be found in `dbrda` and therefore we cannot reconstruct input data from their components. However, if we extract **Y**, **X** and **Z** from similar `rda`, we can reconstruct `dbrda` input **YY'** (you can try if you do not believe).  Because of this, `stressplot` cannot be implemented for partial models or models with more than one component of variation (since distances cannot be reconstucted). `simulate` is also impossible (and it is currently disabled). Some other functions require re-thinking and testing. However, `anova` gives point-to-point same results with the same permutations with all three (which means that `varpart` works as well as far as I understand). Components of variation are OK, but the reconstruction of input distances fails.

**Comment by jarioksa** [comment]
Issue #254 has more about amending `dbrda` results with species scores.

**Comment by jarioksa** [comment]
I have decided to keep both `capscale` and `dbrda` in **vegan**. This decision can be changed later.

The main reason is that `dbrda` is not only an improved implementation of `capscale`, but `dbrda` cannot do some things that we can do with `capscale`. Most importantly, the components of variation are not additive in `dbrda` with consequences to several support functions. See the comments above.
